### DISCUSSION TITLE
Can I use Profiles & Migrations on the same class?

### DISCUSSION THREAD

**Discussion by roa-nyx** [fluentmigrator/fluentmigrator#1564] [Q&A] [answered]
Migrations appear to always be included regardless of profiles.

## Setup

I have two migrations:

```
[Migration(1, "Initial database setup")]
public class Initial : AutoReversingMigration
{
    //...
}
```

```
//To test things out I set the Profile value to something I would never use "Blue".
[Profile("Blue")]
[Migration(2, "Initial database test data")]
public class InitialTestData : Migration
{
    //...
}
```

## Runner

```
var serviceProvider = new ServiceCollection()
	.AddLogging(lb => lb.AddFluentMigratorConsole())
	.AddFluentMigratorCore()
	.ConfigureRunner(rb => rb
		.WithVersionTable(new FluentMigratorVersionTableMetadata())
		.AddSqlServer()
		.WithGlobalConnectionString(Settings.GetMigratorConnectionString())
		.ScanIn(Assembly.GetExecutingAssembly()).For.Migrations()
	)
	.Configure<RunnerOptions>(ro => ro.Profile = Settings.Debug ? "Development" : "Production")
	.BuildServiceProvider();
```

```
serviceProvider.GetRequiredService<IMigrationRunner>().MigrateUp()
```

I would expect the "Blue" migration to never run. Instead it runs whenever I trigger the runner.

 - OS: Windows 10 (fully up to date)
 - Platform .Net Core 6
 - FluentMigrator version 3.3.1
 - FluentMigrator runner, in-process runner
 - Database Management System SQLServer
 - Database Management System Version Microsoft SQL Server Web (64-bit) 14.0.3023.8

I also tried to replace `.ScanIn(Assembly.GetExecutingAssembly()).For.Migrations()` with `.WithMigrationsIn(Assembly.GetExecutingAssembly())` and `.AddSingleton<IAssemblySourceItem>(sp => new AssemblySourceItem(Assembly.GetExecutingAssembly()))`.

Nothing worked.

**Comment by jzabroski** [comment, marked as answer]
It looks like you are using Profiles, when you probably mean to use Tags.

I'm not sure why it would run the Blue Profile, but it is unusual to see someone apply a ProfileAttribute when a MigrationAttribute is also applied.  At the very least, that is not something I have ever done, and I don't think it makes a ton of sense.

Also, I recommend checking out MaintenanceMigrations instead of Profiles.  To be honest, I dont use Profiles any more, since they are basically a more limited version of maintenance migrations.

EDIT: I moved this to a discussion, and retitled your issue into a Q&A type title.  Please let me know if you're all set.

**Reply by roa-nyx** [reply]
Ok, interesting. I thought I was emulating what I found in the docs for doing a kind of "conditional migration" type thing. I'll give MaintenanceMigrations a look.

**Reply by jzabroski** [reply]
There is a sample for Conditional migrations but I don't love Conditionals, on account of the fact that it loads the version table only once per run, as opposed to once per migration.  Thus, it cannot informally reason about the transitive closure of all operations that occur.
